### PR TITLE
CRM-18397: Blank billing address created on paid event registrations with no billing address

### PR DESCRIPTION
**Pull request by rohankatkar**
https://issues.civicrm.org/jira/browse/CRM-18397

### CODE REVIEW

**Comment by rohankatkar** [comment]
jenkin, please retest this

**Comment by rohankatkar** [comment]
jenkins, retest this please

**Comment by rohankatkar** [comment]
@JKingsnorth, following cases are covered in this PR:

1) Event registration with 'external' payment processor.
2) Event registration with 'pay_later' and 'isBillingAddressRequiredForPayLater'.

**Comment by JKingsnorth** [comment]
Manual testing does not pass, commented on JIRA.

**Comment by rohankatkar** [comment]
@JKingsnorth , I agree with your comment which you've mentioned on JIRA. If you changed any information for contact it does not update.

This PR changes only fixes the following scenario:
1) If you made a payment via external payment processor (here I am creating test contribution via 'PayPal Pro'), then it does not ask to fill billing address field as it redirects to the pay pal. After creating contribution, there is no blank billing address is added on contact summary page (no billing address section on contact summary page).

2) As mentioned the case #2 above in my comment, we also need to handle the case if 'pay_later' and 'isBillingAddressRequiredForPayLater' is set to true. In this case, if you changed any information for contact then, it gets updated. And of course 'Billing Address' is added for that contact.

I have also tested with your PR (https://github.com/civicrm/civicrm-core/pull/8122). As you mentioned in issue description,  if we don't ask for billing details on the form, then we don't want to store them in Civi. So don't create a blank billing address.

In your PR, if I create contribution via external payment processor, then its added blank 'Billing Address' on contact summary page only with billing names and empty address fields. I think this is odd.

So,  I think as you mentioned in comment on JIRA, need to handle changes to updates contact information.

**Comment by JKingsnorth** [comment]
I'll take another look at the other PR. On my last testing it didn't create the blank billing details, but I'll re-check with a couple of external providers to see what the problem is - thanks for the feedback.

I'll also take a look at it in the context of 'pay later' with the 'billing address required' setting that you mention.

I'll take a look at this as soon as I can tomorrow. Thanks again.

**Comment by rohankatkar** [comment]
@JKingsnorth, I have updated a PR for following scenarios:

1) Event registration with 'external' payment processor.
2) If an authenticated user makes the contribution then, contact fields are getting updated (e.g Email Address, Contact name).
3) If an anonymous user registers for an event then, a new contact is created.
4) If 'pay_later' and 'isBillingAddressRequiredForPayLater' is set to true, then, billing address is updated or added for that contact.

can you please check if PR satisfies above scenarios?

File: CRM/Event/Form/Registration/Confirm.php [hunk 1]
```diff
@@ -493,6 +493,28 @@ public function postProcess() {
...
+        foreach ($addressFields as $key) {
+          unset($value[$key]);
+        }
+        unset($value['billing_first_name']);
+        unset($value['billing_middle_name']);
+        unset($value['billing_last_name']);
+        unset($value["billing_country_id-{$this->_bltID}"]);
+        unset($value["address_name-{$this->_bltID}"]);
+      }
+
```
Comment (monishdeb) [inline CRM/Event/Form/Registration/Confirm.php]: It looks redundant as the above code 469-494 does the same job - unset billing address field (should have extended the above IF condition to do the same). Also looks like we are simply wiping out the billing address without a criteria.

Comment (JKingsnorth) [inline CRM/Event/Form/Registration/Confirm.php]: I also have a bit of a vendetta against 'lists of fields' in Civi that are manually typed out like this. They are quite hard to maintain. I'll take another look at this in the alternate PR.

**Comment by rohankatkar** [comment]
@JKingsnorth, I am using  'PayPal - Website Payments Pro' processor for testing. 
I think, statndard solution would be just do not create billing address for contact, if billing address fields are empty.

I am closing this PR, because it may affect other use-cases where billing address is required in case of other external payment processor. you can cherry-pick my commit if its helpful.

**Comment by magnolia61** [comment]
Am I correct this PR is not merged? Any chance it will be?

**Comment by JKingsnorth** [comment]
@magnolia61 Would you be able to test the other PR we've proposed? https://github.com/civicrm/civicrm-core/pull/8122 If it works for you then that would be a good bit of review. If you find any issues then do let us know on the issue on JIRA, or on that PR.

**Pull request closed without merging**